adr/00-packages.md

```diff
@@ -0,0 +1,103 @@
+# Architecture Decision Record: Package structure
+
+## Context
+
+We need to specify structure for the application. 
+It will involve choosing how we will divide code into packages and modules. 
+In its current state, application resembles an MVC based architecture. 
+We would like to make a clear decision, what application structure should be going forward.
+
+## Decision
+
+We will divide application into layers represented by packages.
+
+We decided on the following layers:
+* app
+* api
+* infrastructure
+
+The app layer will contain domain models and service interfaces. The api layer will contain endpoint descriptions, without logic.
+Infrastructure layer will house concrete implementations of infrastructure-specific elements. 
+These include database queries based on concrete DB and access library (Quill), and, in the future, possible other technical components like external services, cache, etc. 
+For example, User model, UserRepository interface and UserService interface would be placed in the app layer.
+Postgres based implementation of UserRepository would be placed in the infrastructure layer.  
+  
+Example with one implementation per service(no separate app and infrastructure package):
+
+```
+common
+├── BaseEndpoints
+├── Configuration
+├── Pagination
+├── ErrorInfo
+└── Exceptions
+auth
+└── AuthService
+articles
+├── api
+│   ├── ArticleResponse
+│   ├── ArticlesListResponse
+│   ├── ArticleCreateData
+│   ├── ArticleCreateRequest
+│   ├── ArticleUpdateData
+│   ├── ArticleUpdateRequest
+│   └── ArticlesEndpoints
+├── Author
+├── Article
+├── ArticlesRepository
+├── ArticlesService
+├── Tag
+├── TagsRepository
+├── TagsService
+└── ArticlesServerEndpoints
```

Comment:
what's the difference between `ArticlesServerEndpoints` and `ArticlesEndpoints`?

Reply:
are you want to separate endpoint descriptions from their implementations? is there any functional benefit? otherwise I guess we can keep them combined

Reply:
@adamw That's an approach we took in the Academy. We wanted to keep a standard of separating "the contract" from the implementation. One reason is readability - you can browse pure endpoint declarations as a documentation of the API. Back then I also thought that opening a possibility to publish the contract as a .jar might open some possibilities, like generating clients or sharing value classes, but maybe it's an overkill idea, idk.